### PR TITLE
CPR-500 Refactor of unmerge service

### PR DESCRIPTION
**Pull request by H-Iwanejko**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/jpa/entity/PersonEntity.kt [hunk 1]
```diff
@@ -118,6 +123,21 @@ class PersonEntity(
...
 
+  fun addExcludeOverrideMarker(excludeRecord: PersonEntity) {
+    this.overrideMarkers.add(
+      OverrideMarkerEntity(markerType = OverrideMarkerType.EXCLUDE, markerValue = excludeRecord.id, person = this),
+    )
+  }
+
+  fun removeMergedLink() {
+    this.mergedTo = null
+  }
```
Comment (H-Iwanejko) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/jpa/entity/PersonEntity.kt:134]: Lazy func here, but it makes the calling code a bit more readable

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt [hunk 2]
```diff
@@ -128,9 +128,6 @@ class ReclusterService(
     // Step 1: Mark the source key as merged
     sourcePersonKey.mergedTo = targetPersonKey.id
     sourcePersonKey.status = UUIDStatusType.RECLUSTER_MERGE
-
-    // Step 2: Detach source entity from its original personKey
-    sourcePersonKey.personEntities.remove(sourcePersonEntity)
```
Comment (H-Iwanejko) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt:133]: Not needed as handled by hibernate

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/UnmergeService.kt [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      personRepository.save(reactivatedPersonEntity),
+      personRepository.save(unmergedPersonEntity),
+    )
+  }
+
+  private fun setClusterToNeedsAttentionIfAdditionalRecords(unmergedPersonEntity: PersonEntity, reactivatedPersonEntity: PersonEntity) {
+    when {
+      clusterContainsAdditionalRecords(unmergedPersonEntity, reactivatedPersonEntity) -> {
+        // Get latest person key from persistence
+        val personKey = personKeyRepository.findByPersonId(unmergedPersonEntity.personKey!!.personId)
```
Comment (H-Iwanejko) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/UnmergeService.kt:121]: Needed as trying to update through entity causes transaction error

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/CreateUpdateService.kt [hunk 4]
```diff
@@ -26,40 +27,40 @@ import uk.gov.justice.digital.hmpps.personrecord.service.type.TelemetryEventType
...
   }
 
-  private fun processPerson(person: Person, event: String?, linkRecord: Boolean, callback: () -> PersonEntity?): PersonEntity {
-    val existingPersonEntity: PersonEntity? = callback()
-    val personEntity: PersonEntity = when {
-      (existingPersonEntity == null) -> handlePersonCreation(person, event, linkRecord)
-      else -> handlePersonUpdate(person, existingPersonEntity, event)
-    }
-    return personEntity
+  @Transactional
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/CreateUpdateService.kt:44]: https://docs.spring.io/spring-framework/reference/data-access/transaction/declarative/annotations.html
This doesn't do anything - @Transactional is only effective on methods invoked from other classes

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/DeletionService.kt [hunk 5]
```diff
@@ -33,6 +34,7 @@ class DeletionService(
     }
   }
 
+  @Transactional
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/DeletionService.kt:37]: as above - ineffective on private methods

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt [hunk 6]
```diff
@@ -19,27 +20,28 @@ import uk.gov.justice.digital.hmpps.personrecord.service.TelemetryService
...
 ) {
 
-  @Transactional
   fun processMerge(mergeEvent: MergeEvent, sourcePersonCallback: () -> PersonEntity?, targetPersonCallback: () -> PersonEntity?) = runBlocking {
     runWithRetry(MAX_ATTEMPTS, retryDelay, ENTITY_RETRY_EXCEPTIONS) {
       processMergingOfRecords(mergeEvent, sourcePersonCallback, targetPersonCallback)
     }
   }
 
+  @Transactional
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt:38]: as above

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt [hunk 7]
```diff
@@ -19,27 +20,28 @@ import uk.gov.justice.digital.hmpps.personrecord.service.TelemetryService
 import uk.gov.justice.digital.hmpps.personrecord.service.type.TelemetryEventType.CPR_MERGE_RECORD_NOT_FOUND
 import uk.gov.justice.digital.hmpps.personrecord.service.type.TelemetryEventType.CPR_RECORD_MERGED
 
-@Service
+@Component
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt:23]: @Component and @Service do the same thing anyway. Shall we just go for @Component throughout?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt [hunk 8]
```diff
@@ -19,27 +20,28 @@ import uk.gov.justice.digital.hmpps.personrecord.service.TelemetryService
...
   }
 
+  @Transactional
   private suspend fun processMergingOfRecords(mergeEvent: MergeEvent, sourcePersonCallback: () -> PersonEntity?, targetPersonCallback: () -> PersonEntity?) {
     val (sourcePersonEntity, targetPersonEntity) = collectPeople(sourcePersonCallback, targetPersonCallback)
 
     when {
-      targetPersonEntity == null -> handleTargetRecordNotFound(mergeEvent)
-      sourcePersonEntity == null -> handleSourceRecordNotFound(mergeEvent, targetPersonEntity)
+      targetPersonEntity == PersonEntity.empty -> handleTargetRecordNotFound(mergeEvent)
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt:43]: quite nice, could it be done as an extension function on PersonEntity?

Comment (markreesmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt:43]: I wish I knew more Kotlin

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt [hunk 9]
```diff
@@ -105,15 +108,22 @@ class MergeService(
         EventKeys.SOURCE_SYSTEM to mergeEvent.mergedRecord.sourceSystemType.name,
       ),
     )
-    mergeRecord(mergeEvent, null, targetPersonEntity) { _, targetPerson ->
-      updateTargetRecord(mergeEvent, targetPerson)
+    mergeRecord(mergeEvent, PersonEntity.empty, targetPersonEntity) { _, targetPerson ->
```
[the remark this answers is not in the document]
Comment (markreesmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt:111]: although less confusing than what's on the left

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt [hunk 10]
```diff
@@ -43,6 +41,7 @@ class ReclusterService(
     }
   }
 
+  @Transactional
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt:44]: as above

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt [hunk 11]
```diff
@@ -128,9 +128,6 @@ class ReclusterService(
     // Step 1: Mark the source key as merged
     sourcePersonKey.mergedTo = targetPersonKey.id
     sourcePersonKey.status = UUIDStatusType.RECLUSTER_MERGE
-
-    // Step 2: Detach source entity from its original personKey
-    sourcePersonKey.personEntities.remove(sourcePersonEntity)
     personKeyRepository.save(sourcePersonKey)
 
     // Step 3: Attach the source entity to the target personKey
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt:133]: not mad on the comments but this is no longer step 3 (which is a good illustration of what I don't like about comments - they get out of date quickly)

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/UnmergeService.kt [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+  @Value("\${retry.delay}") private val retryDelay: Long,
+) {
+
+  fun processUnmerge(unmergeEvent: UnmergeEvent, reactivatedPersonCallback: () -> PersonEntity?, unmergedPersonCallback: () -> PersonEntity?) = runBlocking {
+    runWithRetry(MAX_ATTEMPTS, retryDelay, ENTITY_RETRY_EXCEPTIONS) {
+      processUnmergingOfRecords(unmergeEvent, reactivatedPersonCallback, unmergedPersonCallback)
+    }
+  }
+
+  @Transactional
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/UnmergeService.kt:38]: as above

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/PersonKeyService.kt [hunk 13]
```diff
@@ -39,7 +33,7 @@ class PersonKeyService(
       personEntity,
       mapOf(EventKeys.UUID to personKey.personId.toString()),
     )
-    return personKeyRepository.saveAndFlush(personKey)
+    return personKeyRepository.save(personKey)
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/person/PersonKeyService.kt:36]: it is a suprising side effect of a method called `getXX` that it should persist. I see that it already did this

**Review by markreesmoj — COMMENTED** [review]
💬 will have another look later

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt (hunk 8 above)
Comment (H-Iwanejko) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/MergeService.kt:41]: Yeah i thought about this, but the extension func does not mark the entity as have been null check and still need casting `!!`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/api/controller/SearchController.kt [hunk 14]
```diff
@@ -111,12 +111,6 @@ class SearchController(
...
 
   private fun buildIdentifierRecord(personEntity: PersonEntity): PersonIdentifierRecord? {
-    return when (personEntity.sourceSystem) {
-      SourceSystemType.DELIUS -> PersonIdentifierRecord(id = personEntity.crn!!, SourceSystemType.DELIUS.name)
-      SourceSystemType.NOMIS -> PersonIdentifierRecord(id = personEntity.prisonNumber!!, SourceSystemType.NOMIS.name)
-      SourceSystemType.COMMON_PLATFORM -> PersonIdentifierRecord(id = personEntity.defendantId!!, SourceSystemType.COMMON_PLATFORM.name)
-      SourceSystemType.LIBRA -> PersonIdentifierRecord(id = personEntity.defendantId!!, SourceSystemType.LIBRA.name)
-      else -> null
-    }
+    return personEntity.extractSourceSystemId()?.let { PersonIdentifierRecord(id = it, personEntity.sourceSystem.name) }
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/api/controller/SearchController.kt:114]: Could this return a pair or a small object containing the source system ID and the identifier? As far as I can tell from a quick glance we always use both after calling this method

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/jpa/entity/PersonKeyEntity.kt [hunk 15]
```diff
@@ -40,6 +40,7 @@ class PersonKeyEntity(
   var version: Int = 0,
 
 ) {
+
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/jpa/entity/PersonKeyEntity.kt:43]: ❤️

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt (hunk 10 above)
Comment (markreesmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt:44]: this annotation needs moving to the class or a public method, or removing altogether

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt [hunk 16]
```diff
@@ -125,19 +125,13 @@ class ReclusterService(
...
-
-    // Step 2: Detach source entity from its original personKey
-    sourcePersonKey.personEntities.remove(sourcePersonEntity)
     personKeyRepository.save(sourcePersonKey)
 
-    // Step 3: Attach the source entity to the target personKey
     sourcePersonEntity.personKey = targetPersonKey
     personRepository.save(sourcePersonEntity)
 
-    // Step 4: Add the source entity to the target's personEntities list
```
Comment (markreesmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/service/message/ReclusterService.kt:140]: ❤️

**Review by markreesmoj — COMMENTED** [review]
💬

File: src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/api/controller/SearchController.kt (hunk 14 above)
Comment (H-Iwanejko) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/personrecord/api/controller/SearchController.kt:114]: I think it is some what redundant as we would be extracting that from the record which already has the correct `source system` so no extra logic to figure it out. Just need to know what identifier to return.

**Review by markreesmoj — APPROVED** [review]
I think this is OK - it's pretty big and I don't understand everything it does - but OK

**Pull request merged**